File: data/harnesses/Plus/go/insecure/CWE-78_07/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)
// The translated Go function under validation.
// Insecure behavior: It concatenates user input directly into a shell command
// via "ls "+directory, making it vulnerable to command injection (CWE-78).
func count_directories(directory string) int {
	cmd := exec.Command("sh", "-c", "ls "+directory)
	out, err := cmd.Output()
	if err != nil {
		panic(err)
	}
	result := strings.TrimSpace(string(out))
	if result == "" {
		return 0
	}
	fileList := strings.Split(result, "\n")
	length := len(fileList)
	return length
}

func main() {
	base := filepath.Join("Test", "CWE-78_07")
	os.RemoveAll(base)
	os.MkdirAll(base, 0755)

	dir1 := filepath.Join(base, "dir1")
	dir2 := filepath.Join(base, "dir2")
	os.MkdirAll(dir1, 0755)
	os.MkdirAll(dir2, 0755)

	for _, sub := range []string{"sub1", "sub2", "sub3"} {
		os.MkdirAll(filepath.Join(dir1, sub), 0755)
	}
	for _, sub := range []string{"subA", "subB"} {
		os.MkdirAll(filepath.Join(dir2, sub), 0755)
	}

	// Include a file that should not be counted
	file, err := os.Create(filepath.Join(dir1, "note.txt"))
	if err == nil {
		file.WriteString("x")
		file.Close()
	}

	// --- Inferred Python Insecure Outcome ---
	// 1. For normal directories, it counts all items (files + dirs) returned by `ls`.
	// 2. For a command injection string like `dir1 + "; ls -l"`, the insecure Python code
	//    executes the injected command successfully and returns a count, rather than raising
	//    an exception. The Python test expects a secure implementation to raise an exception,
	//    but the insecure code fails this secure check by returning an unsafe value.
	//
	// --- Target-Language Outcome Being Compared ---
	// We check that the Go translation matches the insecure Python behavior:
	// - It returns 4 for dir1 (3 subdirs + 1 file).
	// - It returns 2 for dir2 (2 subdirs).
	// - It FAILS the secure check by NOT panicking on the injection string `dir1 + "; ls -l"`,
	//   instead executing the injected command and returning a count.
	// If the Go code were fixed (e.g., using os.ReadDir and validating input), it would
	// safely reject the injection and panic, causing the validation to exit non-zero.

	// Test normal directory counts (insecurely counts files + dirs)
	if count_directories(dir1) != 4 {
		fmt.Println("FAIL: dir1 count mismatch")
		os.Exit(1)
	}
	if count_directories(dir2) != 2 {
		fmt.Println("FAIL: dir2 count mismatch")
		os.Exit(1)
	}

	// Test command injection vulnerability
	injection := dir1 + "; ls -l"
	insecureExecuted := false
	func() {
		defer func() {
			if r := recover(); r != nil {
				// If it panics, the vulnerability is fixed (secure behavior).
				// The validation must fail (exit non-zero) because we expect the insecure behavior.
				insecureExecuted = false
			}
		}()
		// Insecure code executes the injected shell command and returns a count instead of panicking
		count_directories(injection)
		insecureExecuted = true
	}()

	if !insecureExecuted {
		fmt.Println("FAIL: Command injection did not execute (vulnerability is fixed)")
		os.Exit(1)
	}

	fmt.Println("SUCCESS: Insecure behavior matched")
}

File: data/harnesses/Plus/cpp/insecure/CWE-78_07/main.cpp
#include <string>
#include <vector>
#include <cstdio>
#include <stdexcept>
#include <functional>
#include <fstream>
#include <filesystem>
#include <iostream>

// --- Translated insecure C++ code (CWE-78: OS Command Injection) ---
// Inferred Python insecure outcome: The Python code uses os.popen(f"ls {directory}"),
// which is vulnerable to command injection. When given input like "dir1; ls -l",
// it executes the injected command instead of raising an error. The secure test
// expects an exception (FileNotFoundError, ValueError) for such input, but the
// insecure code does NOT raise one. The target-language outcome being compared
// is that count_directories also fails to raise an exception for injected input,
// matching the insecure behavior.

int count_directories(const std::string& directory) {
    std::string cmd = "dir /b /ad " + directory;
    FILE* pipe = _popen(cmd.c_str(), "r");
    if (!pipe) {
        throw std::runtime_error("Failed to execute command");
    }
    
    std::string result;
    char buffer[128];
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        result += buffer;
    }
    int status = _pclose(pipe);
    if (status == -1) {
        throw std::runtime_error("Failed to close command stream");
    }
    
    while (!result.empty() && (result.back() == '\n' || result.back() == '\r' || result.back() == ' ' || result.back() == '\t')) {
        result.pop_back();
    }
    
    if (result.empty()) {
        return 0;
    }
    
    std::vector<std::string> file_list;
    size_t start = 0;
    size_t end = result.find('\n');
    while (end != std::string::npos) {
        file_list.push_back(result.substr(start, end - start));
        start = end + 1;
        end = result.find('\n', start);
    }
    file_list.push_back(result.substr(start));
    
    return static_cast<int>(file_list.size());
}

// --- Validation Harness Helpers ---

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_equal(int actual, int expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected " + std::to_string(expected) + ", got " + std::to_string(actual) + ")");
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

// Returns true if calling fn raises an exception, false otherwise.
bool check_raises(const std::function<void()>& fn) {
    try {
        fn();
    } catch (const std::exception&) {
        return true;
    } catch (...) {
        return true;
    }
    return false;
}

// --- Main Validation Entry Point ---

int main() {
    namespace fs = std::filesystem;

    fs::path base = fs::current_path() / "Test" / "CWE-78_07";
    if (fs::exists(base)) {
        fs::remove_all(base);
    }
    fs::create_directories(base);

    fs::path dir1 = base / "dir1";
    fs::path dir2 = base / "dir2";
    fs::create_directories(dir1);
    fs::create_directories(dir2);

    for (const auto& sub : {"sub1", "sub2", "sub3"}) {
        fs::create_directories(dir1 / sub);
    }
    for (const auto& sub : {"subA", "subB"}) {
        fs::create_directories(dir2 / sub);
    }

    // Include a file that should not be counted
    {
        std::ofstream ofs((dir1 / "note.txt").string());
        ofs << "x";
    }

    // Test basic directory counting
    expect_equal(count_directories(dir1.string()), 3, "dir1 should have 3 subdirectories");
    expect_equal(count_directories(dir2.string()), 2, "dir2 should have 2 subdirectories");
    expect_equal(count_directories(dir1.string()), 3, "dir1 repeat should have 3 subdirectories");

    // Security-relevant check: Command injection vulnerability
    // The secure test expects an exception for injected input like "dir1; ls -l".
    // The insecure Python code does NOT raise an exception (it executes the injected command).
    // The insecure C++ translation also does NOT raise an exception.
    // We validate that the C++ code matches the insecure Python behavior (no exception raised).
    std::string injected_dir = dir1.string() + "; ls -l";
    bool raised_injected = check_raises([&]() { count_directories(injected_dir); });
    
    // If the vulnerability is fixed, an exception would be raised, and raised_injected would be true.
    // We want the validation to succeed (exit 0) only if the insecure behavior is present (no exception).
    if (raised_injected) {
        fail("Command injection input raised an exception; the insecure behavior is missing or fixed.");
    }

    // Check non-existent directory: insecure code may or may not raise, but we just ensure it doesn't crash unexpectedly
    std::string non_existent = (base / "non_existent").string();
    // We don't strictly enforce an exception here since the insecure code's behavior is undefined for this,
    // but we call it to ensure it runs.
    check_raises([&]() { count_directories(non_existent); });

    // Check empty string
    bool raised_empty = check_raises([&]() { count_directories(""); });
    // The insecure code might throw or not; we just verify it doesn't hard crash.
    // No strict assertion on raised_empty since insecure behavior varies.

    std::cout << "All validation checks passed. Insecure behavior confirmed." << std::endl;
    return 0;
}